### PR TITLE
[IGNITE-19496] Sql. Implement SqlSchemaManager on top of CatalogService.

### PR DESCRIPTION
**Pull request by lowka**
Provides implementation of SqlSchemaManager backed by catalog  + removes constraints that require primary key columns from TableDescriptor constructor.

Adds Implementation of IgniteTable (`IgniteSchemaTable`) that supports only planning related operations (with execution related methods throwing exceptions). Index related methods are moved to `IgniteSchemaIndex`.

### CODE REVIEW

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/TableDescriptor.java [hunk 1]
```diff
@@ -88,6 +86,18 @@ public int engineId() {
         return engineId;
     }
 
+    public TableColumnDescriptor[] columns() {
+        return columns;
+    }
+
+    public boolean isKeyColumn(String name) {
+        return Arrays.binarySearch(primaryKeyColumns, name) >= 0;
```
Comment (AMashenkov) [inline modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/TableDescriptor.java:94]: primaryKeyColumns array is not sorted.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java [hunk 2]
```diff
@@ -483,4 +486,147 @@ public static RelDataType native2relationalType(RelDataTypeFactory factory, Nati
...
+                //fallthrough
+            case NUMBER:
+            case NULL:
+            default:
+                throw new IllegalArgumentException("Type is not supported: " + columnType);
+        }
+    }
+
+    /** Converts provides {@link RelDataType} to {@link NativeType}.*/
+    public static NativeType relType2NativeType(RelDataType dataType) {
```
Comment (lowka) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/TypeUtils.java:542]: Looks like a similar function exists in IgniteTypeFactory and that function should be used instead.

File: modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/TableDescriptor.java (hunk 1 above)
Comment (lowka) [reply modules/catalog/src/main/java/org/apache/ignite/internal/catalog/descriptors/TableDescriptor.java:94]: @AMashenkov I think there similar methods for checking whether in your or Pavel's patches and I am going to use those  after you patches get merged.

**Review by korlov42 — COMMENTED** [review]
@lowka, the patch is enormous! Let's derive all TestCatalog-related refactoring to a separate ticket, leaving here only implementation of SchemaManager over catalog with necessary self tests

**Comment by lowka** [comment]
Updated PR, changes related to integration of a catalog backed schema manager has been removed.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java [hunk 3]
```diff
@@ -254,6 +255,8 @@ private CompletableFuture<QueryPlan> prepareDml(SqlNode sqlNode, PlanningContext
             // Convert to Relational operators graph
             IgniteRel igniteRel = optimize(validatedNode, planner);
 
+            System.err.println(RelOptUtil.dumpPlan("DUMP", igniteRel, SqlExplainFormat.TEXT, SqlExplainLevel.NON_COST_ATTRIBUTES));
```
Comment (AMashenkov) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java:258]: Let's revert this.

Comment (lowka) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java]: Sure.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/IgniteIndex.java [hunk 4]
```diff
@@ -82,12 +84,22 @@ public enum Type {
      */
     public IgniteIndex(Index<?> index) {
         this.index = Objects.requireNonNull(index, "index");
```
Comment (lowka) [inline modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/IgniteIndex.java:86]: I think I should revert these changes as they are not necessary.

**Comment by lowka** [comment]
Resolved conflicts.

**Comment by lowka** [comment]
After discussion with Pavel, I updated sql schema manager to use length field.
@AMashenkov Could you please take a look?

**Review by AMashenkov — APPROVED** [review]

**Review by AMashenkov — APPROVED** [review]
